File: outputDir/vscripts/features/OperatorsDemo.lua
-- package: com.example.features

require("ktox-lib")
ktox_sourcemap_traceback(debug and debug.getinfo and (debug.getinfo(1) or {}).short_src or "", "features/OperatorsDemo.kt", {["1-15"]=1,["16"]=13,["17"]=14,["18"]=15,["19"]=16,["20-23"]=17,["24"]=22,["25"]=23,["26"]=24,["27"]=25,["28-31"]=26,["32"]=31,["33"]=32,["34"]=33,["35"]=34,["36-39"]=35,["40"]=40,["41"]=41,["42"]=42,["43"]=43,["44-47"]=44,["48"]=49,["49"]=50,["50"]=51,["51"]=52,["52-55"]=53,["56"]=58,["57"]=59,["58"]=60,["59"]=61,["60-63"]=62,["64"]=68,["65"]=69,["66"]=70,["67"]=71,["68-71"]=72,["72"]=77,["73"]=78,["74-77"]=79,["78"]=84,["79"]=85,["80-81"]=86,["82-85"]=88,["86"]=93,["87"]=94,["88"]=95,["89"]=96,["90-92"]=97,["93-96"]=100,["97"]=105,["98"]=106,["99"]=107,["100-101"]=108,["102-105"]=110,["106"]=115,["107"]=116,["108"]=117,["109"]=118,["110-113"]=119,["114"]=124,["115"]=125,["116"]=126,["117-120"]=127,["121"]=132,["122"]=133,["123"]=134,["124-125"]=135,["126-128"]=137})

OperatorsDemo = {}
OperatorsDemo.__index = OperatorsDemo

function OperatorsDemo:new()
    local self = setmetatable({}, OperatorsDemo)
    return self
end


function demoCompoundAssignment()
    local x = 10
    x = ktox_plusAssign(x, 5)
    x = ktox_minusAssign(x, 3)
    x = x * 2
    return x
end

function demoDivMod()
    local a = 20
    a = a / 4
    local b = 17
    b = b % 5
    return a + b
end

function demoPostIncDec()
    local a = 5
    a = a + 1
    a = a + 1
    a = a - 1
    return a
end

function demoPreIncDec()
    local b = 0
    b = b + 1
    b = b + 1
    b = b - 1
    return b
end

function demoUnary()
    local pos = 7
    local neg = -pos
    local t = true
    local f = not t
    return (not f and neg + 10 or 0)
end

function demoBooleans()
    local tt = true and true
    local tf = true and false
    local ft = false or true
    local ff = false or false
    return ((tt and 1 or 0)) + ((tf and 0 or 1)) + ((ft and 1 or 0)) + ((ff and 0 or 1))
end

function demoInequality()
    local x = 5
    local y = 3
    local ne = x ~= y
    local eq = x ~= x
    return ((ne and 1 or 0)) + ((eq and 0 or 1))
end

function demoModulo()
    local a = 17 % 5
    local b = 10 % 3
    return a + b
end

function demoDoWhile()
    local n = 0
    repeat
        n = n + 1
    until not (n < 5)
    return n
end

function demoBreak()
    local found = 0
    for i = 1, 10 do
        if i == 4 then
            found = i
            break
        end
    end
    return found
end

function demoForList()
    local items = {10, 20, 30, 40}
    local total = 0
    for _, item in ipairs(items) do
        total = ktox_plusAssign(total, item)
    end
    return total
end

function demoElvisStandalone()
    local a = nil
    local b = 42
    local ra = ktox_elvis(a, -1)
    local rb = ktox_elvis(b, -1)
    return ra + rb
end

function demoInRangeExpression()
    local x = 5
    local inRange = (x >= 1 and x <= 10)
    local notIn = not (x >= 1 and x <= 4)
    return ((inRange and 1 or 0)) + ((notIn and 1 or 0))
end

function demoDestructuringFor()
    local scores = {["a"] = 10, ["b"] = 20, ["c"] = 30}
    local total = 0
    for k, v in pairs(scores) do
        total = ktox_plusAssign(total, v)
    end
    return total
end

File: src/main/kotlin/com/example/features/OperatorsDemo.kt
package com.example.features

/**
 * Demonstrates operators and control-flow patterns not covered by other demo files:
 * compound assignment, increment/decrement, unary operators, boolean operators,
 * inequality, modulo, do-while, break, for-over-list, standalone Elvis,
 * in-range expressions, and destructuring for loop.
 */
class OperatorsDemo

/** Compound assignment: +=, -=, *= */
fun demoCompoundAssignment(): Int {
    var x = 10
    x += 5   // 15
    x -= 3   // 12
    x *= 2   // 24
    return x
}

/** Compound assignment: /= and %= */
fun demoDivMod(): Int {
    var a = 20
    a /= 4   // 5 (Kotlin integer division; Lua 5.0)
    var b = 17
    b %= 5   // 2 in both Kotlin and Lua
    return a + b  // 7
}

/** Post-increment and post-decrement in statement position */
fun demoPostIncDec(): Int {
    var a = 5
    a++   // 6
    a++   // 7
    a--   // 6
    return a
}

/** Pre-increment and pre-decrement in statement position */
fun demoPreIncDec(): Int {
    var b = 0
    ++b   // 1
    ++b   // 2
    --b   // 1
    return b
}

/** Unary minus and logical NOT */
fun demoUnary(): Int {
    val pos = 7
    val neg = -pos      // -7
    val t = true
    val f = !t          // false
    return if (!f) neg + 10 else 0   // !false → true → -7 + 10 = 3
}

/** Boolean AND and OR short-circuit */
fun demoBooleans(): Int {
    val tt = true && true    // true
    val tf = true && false   // false
    val ft = false || true   // true
    val ff = false || false  // false
    return (if (tt) 1 else 0) + (if (tf) 0 else 1) + (if (ft) 1 else 0) + (if (ff) 0 else 1)
    // 1 + 1 + 1 + 1 = 4
}

/** Inequality (!=) operator */
fun demoInequality(): Int {
    val x = 5
    val y = 3
    val ne = x != y       // true
    val eq = x != x       // false
    return (if (ne) 1 else 0) + (if (eq) 0 else 1)  // 1 + 1 = 2
}

/** Modulo (%) operator */
fun demoModulo(): Int {
    val a = 17 % 5   // 2
    val b = 10 % 3   // 1
    return a + b     // 3
}

/** do-while loop (transpiles to Lua repeat…until) */
fun demoDoWhile(): Int {
    var n = 0
    do {
        n++
    } while (n < 5)
    return n  // 5
}

/** break inside a for loop */
fun demoBreak(): Int {
    var found = 0
    for (i in 1..10) {
        if (i == 4) {
            found = i
            break
        }
    }
    return found  // 4
}

/** for loop iterating over a list (transpiles to Lua for _, item in ipairs(t)) */
fun demoForList(): Int {
    val items = listOf(10, 20, 30, 40)
    var total = 0
    for (item in items) {
        total += item
    }
    return total  // 100
}

/** Elvis ?: used as a standalone expression (no null-safe chain) */
fun demoElvisStandalone(): Int {
    val a: Int? = null
    val b: Int? = 42
    val ra = a ?: -1   // null  → -1
    val rb = b ?: -1   // 42   → 42
    return ra + rb     // 41
}

/** in / !in on a range, used as a boolean expression */
fun demoInRangeExpression(): Int {
    val x = 5
    val inRange = x in 1..10      // true (5 ≥ 1 and 5 ≤ 10)
    val notIn = x !in 1..4        // true (5 > 4)
    return (if (inRange) 1 else 0) + (if (notIn) 1 else 0)  // 2
}

/** Destructuring for loop over a map: for ((k, v) in map) */
fun demoDestructuringFor(): Int {
    val scores = mapOf("a" to 10, "b" to 20, "c" to 30)
    var total = 0
    for ((k, v) in scores) {
        total += v
    }
    return total  // 60
}
